### ISSUE TITLE
Support locking

### ISSUE THREAD

**Issue by xiii** [ITV/domed-city#14]
In the event of 2 users working in the same TF codebase, we modify the plan / apply methods with a a lock helper method that does:

```
def lock_tf_run
   raise("someone else is playing in that environment") unless there_is_no_lock_from_any_user_in_the_s3_bucket(bucket => "environment bucket")
   create_s3_object(filename => "username.lock",contents => "$timestamp", bucket => "environment bucket"
end

def unlock_tf_run
  delete_s3_object(filename => "username.lock", bucket => "environment bucket"
end
```

```
def create_plan
      command         = "terraform plan -module-depth=1 -refresh=true -out=#{@plan_file} -var-file=params/env.tfvars"
      failure_message = 'something went wrong when creating the TF plan'
      *lock_tf_run*
      execute_command(command, failure_message)
    end
```

```
def apply
      command         = "terraform apply #{@plan_file}"
      failure_message = 'something went wrong when applying the TF plan'
      execute_command(command, failure_message)
      *unlock_tf_run*
    end
```

**Comment by madAndroid** [comment]
One of the my colleagues at ScaleFactory has implemented a locking solution using AWS SimpleDB, which appears to offer a potentially better solution than using file locking - this should be more resilient and suffer fewer issues than file locking. I will investigate adding the same or similar solution once it's deemed stable.

**Comment by madAndroid** [comment]
Locking support has been implemented in #25 - this is currently being tested in core and root TF config.

**Comment by madAndroid** [comment]
Terraform now has native state locking - the implementation of that is in #38
